File: src/C++SIM/Stat/src/TimeVariance.cc
/*
 * Copyright (C) 1994
 *
 * Department of Computing Science,
 * The University,
 * Newcastle upon Tyne,
 * UK.
 */


#include <errno.h>
#include <iostream.h>
#include <fstream.h>
#include <iomanip.h>

#ifndef PROCESS_H_
#  include <ClassLib/Process.h>
#endif

#ifndef ERROR_H_
#  include <Common/Error.h>
#endif

#ifndef TIMEVARIANCE_H_
#  include <Stat/TimeVariance.h>
#endif


TimeVariance::TimeVariance () { TimeVariance::reset(); }

TimeVariance::~TimeVariance () {}

void TimeVariance::reset ()
{
    first = TRUE;
    startTime = currentValue = 0.0;
    stime = total = 0.0;
    Variance::reset();
}

double TimeVariance::timeAverage () const
{
    if (first || (Process::CurrentTime() - startTime) == 0)
        return 0.0;

    return ((total + area())/(Process::CurrentTime() - startTime));
}

void TimeVariance::store (double value)
{
    currentValue = value;
    stime = Process::CurrentTime();
}

void TimeVariance::setValue (double value)
{
    Variance::setValue(value);

    if (!first)
    {
        total += area();
        if (value == currentValue)
	    return;
    }
    else
    {
        first = FALSE;
	startTime = Process::CurrentTime();
    }

    store(value);
}

Boolean TimeVariance::saveState (ofstream& oFile) const
{
    if (!oFile)
	return FALSE;

    oFile << " " << first;
    oFile << " " << startTime << " " << currentValue;
    oFile << " " << stime << " " << total;
    
    return Variance::saveState(oFile);
}

Boolean TimeVariance::saveState (const char* fileName) const
{
    ofstream oFile(fileName, ios::out);
    Boolean result;
    
    if (!oFile)
    {
	error_stream << WARNING << "TimeVariance::saveState - error " << errno
		     << " for file " << fileName << endl;
	return FALSE;
    }
    
    result = saveState(oFile);
    oFile.close();
    return result;
}

Boolean TimeVariance::restoreState (ifstream& iFile)
{
    if (!iFile)
	return FALSE;
    
    iFile >> first;
    iFile >> startTime >> currentValue;
    iFile >> stime, total;
    
    return TRUE;
}

Boolean TimeVariance::restoreState (const char* fileName)
{
    ifstream iFile(fileName, ios::in);
    Boolean result;
    
    if (!iFile)
    {
	error_stream << WARNING << "TimeVariance::restoreState - error " << errno
		     << " for file " << fileName << endl;
	return FALSE;
    }
    
    result = restoreState(iFile);
    iFile.close();
    return result;
}

#ifdef NO_INLINES
#  define TIMEVARIANCE_CC_
#  include <Stat/TimeVariance.n>
#  undef TIMEVARIANCE_CC_
#endif

File: src/JavaSim/src/main/java/org/javasim/stats/TimeVariance.java
/*
 * Copyright 1990-2008, Mark Little, University of Newcastle upon Tyne
 * and others contributors as indicated 
 * by the @authors tag. All rights reserved. 
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors. 
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A 
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, 
 * MA  02110-1301, USA.
 * 
 * (C) 1990-2008,
 */

package org.javasim.stats;

import java.io.*;
import java.lang.IllegalArgumentException;

import org.javasim.SimulationProcess;

/**
 * Obtain the average value given during the simulation time.
 */

public class TimeVariance extends Variance
{

    public TimeVariance()
    {
        reset();
    }

    /**
     * Zero the statistics.
     */

    public void reset ()
    {
        first = true;
        startTime = currentValue = 0.0;
        stime = total = 0.0;
        super.reset();
    }

    /**
     * Add 'value', updating the statistics.
     */

    public void setValue (double value) throws IllegalArgumentException
    {
        super.setValue(value);

        if (!first)
        {
            total += area();
            if (value == currentValue)
                return;
        }
        else
        {
            first = false;
            startTime = SimulationProcess.CurrentTime();
        }

        store(value);
    }

    /**
     * Return the average value given up to the current simulation time.
     */

    public double timeAverage ()
    {
        if (first || (SimulationProcess.CurrentTime() - startTime) == 0)
            return 0.0;

        return ((total + area()) / (SimulationProcess.CurrentTime() - startTime));
    }

    /**
     * Save the state of the histogram to the file named 'fileName'.
     */

    public boolean saveState (String fileName) throws IOException
    {
        FileOutputStream f = new FileOutputStream(fileName);
        DataOutputStream oFile = new DataOutputStream(f);

        boolean res = saveState(oFile);

        f.close();

        return res;
    }

    /**
     * Save the state of the histogram to the stream 'oFile'.
     */

    public boolean saveState (DataOutputStream oFile) throws IOException
    {
        oFile.writeBoolean(first);
        oFile.writeDouble(startTime);
        oFile.writeDouble(currentValue);
        oFile.writeDouble(stime);
        oFile.writeDouble(total);

        return super.saveState(oFile);
    }

    /**
     * Restore the histogram state from the file 'fileName'.
     */

    public boolean restoreState (String fileName) throws FileNotFoundException,
            IOException
    {
        FileInputStream f = new FileInputStream(fileName);
        DataInputStream iFile = new DataInputStream(f);

        boolean res = restoreState(iFile);

        f.close();

        return res;
    }

    /**
     * Restore the histogram state from the stream 'iFile'.
     */

    public boolean restoreState (DataInputStream iFile) throws IOException
    {
        first = iFile.readBoolean();
        startTime = iFile.readDouble();
        currentValue = iFile.readDouble();
        stime = iFile.readDouble();
        total = iFile.readDouble();

        return true;
    }

    private double area ()
    {
        return (currentValue * (SimulationProcess.CurrentTime() - stime));
    }

    private void store (double value)
    {
        currentValue = value;
        stime = SimulationProcess.CurrentTime();
    }

    private boolean first;

    private double startTime;

    private double currentValue;

    private double stime;

    private double total;

};
